File: src/song/validate.ts
import { migrateSong } from "./migrate.ts";
import { sectionLength } from "./arrangement.ts";
import { TABLES, type Song, type Pitch } from "./model.ts";
import { add, cmp, time, type Time } from "./time.ts";
import { ok, err, type Result } from "../result.ts";
const assert = (yes: unknown, message: string): void => {
  if (!yes) throw new Error(message);
};
const record = (v: unknown): v is Record<string, unknown> =>
  typeof v === "object" && v !== null && !Array.isArray(v);
const id = (v: unknown) =>
  typeof v === "string" &&
  /^[a-zA-Z0-9_-]{1,100}$/.test(v) &&
  !["__proto__", "constructor", "prototype"].includes(v);
const text = (v: unknown) => typeof v === "string" && v.length <= 10000;
const integer = (v: unknown, min: number, max: number) =>
  typeof v === "number" && Number.isSafeInteger(v) && v >= min && v <= max;
const scalar = (v: unknown, min: number, max: number) =>
  typeof v === "number" && Number.isFinite(v) && v >= min && v <= max;
function duration(v: unknown, positive = false): asserts v is Time {
  assert(
    Array.isArray(v) && v.length === 2,
    "Time must be [numerator, denominator]",
  );
  const a = v as number[];
  const t = time(a[0]!, a[1]!);
  assert(
    cmp(t, [0, 1]) >= (positive ? 1 : 0),
    "Musical time must be nonnegative (durations positive)",
  );
  assert(t[0] === a[0] && t[1] === a[1], "Time fractions must be normalized");
}
function pitch(v: unknown): asserts v is Pitch {
  assert(record(v), "Pitch must be an object");
  const p = v as Pitch;
  assert(
    integer(p.degree, 1, 7) &&
      integer(p.alteration, -4, 4) &&
      integer(p.octave, -5, 5),
    "Invalid relative pitch",
  );
}
export function validateSong(input: unknown): Result<Song> {
  try {
    assert(record(input), "Song must be an object");
    const s = migrateSong(input) as Song;
    assert(s.schemaVersion === 2, "Unsupported song schema version");
    assert(
      id(s.id) &&
        text(s.title) &&
        text(s.mode) &&
        s.degreeReference === "major",
      "Invalid song identity or tonality",
    );
    assert(
      record(s.tempo) && scalar(s.tempo.bpm, 10, 600),
      "Tempo must be between 10 and 600 BPM",
    );
    duration(s.tempo.beatUnit, true);
    assert(record(s.tables), "Song needs entity tables");
    let count = 0;
    for (const table of TABLES) {
      assert(record(s.tables[table]), `Missing table ${table}`);
      for (const [key, e] of Object.entries(s.tables[table])) {
        assert(
          record(e) && id(key) && e.id === key && text(e.name),
          `Invalid identity in ${table}: ${key}`,
        );
        if (++count > 20000) throw new Error("Song exceeds 20,000 entities");
      }
    }
    const t = s.tables;
    const ref = (table: keyof typeof t, v: unknown) =>
      assert(
        typeof v === "string" && Object.hasOwn(t[table], v),
        `Broken ${table} reference: ${String(v)}`,
      );
    for (const p of Object.values(t.parts))
      assert(
        ["guitar", "bass", "drums"].includes(p.instrument) &&
          scalar(p.volume, 0, 1) &&
          typeof p.muted === "boolean",
        "Invalid part",
      );
    for (const v of Object.values(t.voices)) ref("parts", v.partId);
    for (const p of Object.values(t.patterns)) {
      duration(p.length, true);
      assert(cmp(p.length, [10000, 1]) <= 0, "Pattern too long");
      if (p.sourceId !== null) {
        ref("patterns", p.sourceId);
        assert(p.sourceId !== p.id, "Pattern cannot vary itself");
      }
    }
    for (const c of Object.values(t.chords)) {
      assert(
        Array.isArray(c.notes) && c.notes.length > 0 && c.notes.length <= 64,
        "Chord must contain 1–64 notes",
      );
      const ids = new Set<string>();
      for (const n of c.notes) {
        assert(
          record(n) && id(n.id) && !ids.has(n.id),
          "Invalid or duplicate chord member",
        );
        ids.add(n.id);
        pitch(n.pitch);
      }
      assert(c.label === null || text(c.label), "Invalid chord label");
    }
    for (const e of Object.values(t.events)) {
      ref("patterns", e.patternId);
      duration(e.start);
      duration(e.duration, true);
      assert(
        cmp(e.start, t.patterns[e.patternId]!.length) < 0,
        "Event attack must be inside its pattern",
      );
      assert(
        ["note", "chord", "drum", "rest"].includes(e.kind) &&
          ["normal", "staccato", "sustain", "muted", "ghost"].includes(
            e.articulation,
          ),
        "Invalid event type",
      );
      pitch(e.pitch);
      assert(
        ["kick", "snare", "hat"].includes(e.drum) && scalar(e.accent, 0, 1),
        "Invalid event expression",
      );
      if (e.kind === "chord") ref("chords", e.chordId);
      else assert(e.chordId === null, "Only chord events reference chords");
      assert(
        Array.isArray(e.performance) && e.performance.length <= 64,
        "Invalid chord performance",
      );
      assert(
        e.kind === "chord" || e.performance.length === 0,
        "Only chords have member performances",
      );
      const seen = new Set<string>();
      for (const p of e.performance) {
        assert(
          record(p) &&
            !seen.has(p.memberId) &&
            t.chords[e.chordId!]!.notes.some((n) => n.id === p.memberId),
          "Broken/duplicate chord member reference",
        );
        seen.add(p.memberId);
        duration(p.offset);
        duration(p.duration, true);
      }
    }
    for (const b of Object.values(t.bars)) {
      ref("sections", b.sectionId);
      assert(
        integer(b.numerator, 1, 64) &&
          [1, 2, 4, 8, 16, 32, 64].includes(b.denominator),
        "Invalid time signature",
      );
      assert(
        Array.isArray(b.groups) &&
          b.groups.length > 0 &&
          b.groups.every((g) => integer(g, 1, 64)) &&
          b.groups.reduce((a, b) => a + b, 0) === b.numerator,
        "Beat groups must sum to numerator",
      );
      if (b.actual !== null) {
        duration(b.actual, true);
        assert(
          cmp(b.actual, time(b.numerator * 4, b.denominator)) <= 0,
          "Incomplete bar cannot exceed full bar",
        );
      }
    }
    const assigned = new Set<string>();
    for (const sec of Object.values(t.sections)) {
      if (sec.sourceId !== null) {
        ref("sections", sec.sourceId);
        assert(sec.sourceId !== sec.id, "Section cannot vary itself");
      }
      assert(Array.isArray(sec.barIds), "Section needs bar order");
      for (const b of sec.barIds) {
        ref("bars", b);
        assert(
          t.bars[b]!.sectionId === sec.id && !assigned.has(b),
          "Invalid/duplicate section bar",
        );
        assigned.add(b);
      }
    }
    assert(
      assigned.size === Object.keys(t.bars).length,
      "Every bar must appear in its section",
    );
    for (const a of Object.values(t.arrangement)) ref("sections", a.sectionId);
    assert(
      Array.isArray(s.arrangementOrder) &&
        new Set(s.arrangementOrder).size === s.arrangementOrder.length,
      "Invalid arrangement order",
    );
    for (const a of s.arrangementOrder) ref("arrangement", a);
    assert(
      s.arrangementOrder.length === Object.keys(t.arrangement).length,
      "Every section occurrence needs an order",
    );
    for (const o of Object.values(t.occurrences)) {
      if (o.sectionId !== null) ref("sections", o.sectionId);
      ref("patterns", o.patternId);
      ref("voices", o.voiceId);
      duration(o.start);
      duration(o.span, true);
      duration(o.phase);
      if (o.sectionId !== null)
        assert(
          cmp(add(o.start, o.span), sectionLength(s, o.sectionId)) <= 0,
          `Placement ${o.name} exceeds its section; adjust its start/span explicitly`,
        );
      assert(
        cmp(o.phase, t.patterns[o.patternId]!.length) < 0,
        "Phase must be inside pattern",
      );
      assert(
        ["continue", "restart", "stop"].includes(o.boundary) &&
          ["ring", "cut"].includes(o.tails),
        "Invalid boundary choice",
      );
    }
    for (const table of ["phrases", "lyrics"] as const)
      for (const e of Object.values(t[table])) {
        ref("sections", e.sectionId);
        duration(e.start);
        duration(e.duration, true);
        assert(
          cmp(add(e.start, e.duration), sectionLength(s, e.sectionId)) <= 0,
          `${table}/${e.id} exceeds its section`,
        );
      }
    for (const l of Object.values(t.lyrics)) {
      assert(text(l.text), "Invalid lyric text");
      if (l.partId !== null) ref("parts", l.partId);
      if (l.phraseId !== null) {
        ref("phrases", l.phraseId);
        const p = t.phrases[l.phraseId]!;
        assert(
          l.sectionId === p.sectionId &&
            cmp(l.start, p.start) >= 0 &&
            cmp(add(l.start, l.duration), add(p.start, p.duration)) <= 0,
          "Linked phrase must contain its lyric span",
        );
      }
    }
    for (const table of ["sections", "patterns"] as const)
      for (const e of Object.values(t[table])) {
        const seen = new Set([e.id]);
        let parent = e.sourceId;
        while (parent !== null) {
          assert(!seen.has(parent), `Cyclic ${table} lineage`);
          seen.add(parent);
          parent = t[table][parent]!.sourceId;
        }
      }
    for (const m of Object.values(t.markers)) duration(m.at);
    return ok(structuredClone(s));
  } catch (e) {
    return err(e instanceof Error ? e.message : String(e));
  }
}

File: src-tauri/crates/song-core/src/validate.rs
use crate::fretted::validate_fretted;
use crate::{Error, Pitch, Result, Song, Time, ensure, lookup, section_length};
use std::collections::BTreeSet;

pub(crate) fn identity(id: &str) -> bool {
    !id.is_empty()
        && id.len() <= 100
        && id
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || b == b'_' || b == b'-')
        && !["__proto__", "constructor", "prototype"].contains(&id)
}
fn text(s: &str, limit: usize) -> bool {
    s.encode_utf16().count() <= limit
}
fn pitch(p: &Pitch) -> Result<()> {
    ensure(
        (1..=7).contains(&p.degree)
            && (-4..=4).contains(&p.alteration)
            && (-5..=5).contains(&p.octave),
        "Invalid relative pitch",
    )
}
fn positive(t: Time) -> Result<()> {
    ensure(
        t > Time::ZERO,
        "Musical time must be nonnegative (durations positive)",
    )
}
fn nonnegative(t: Time) -> Result<()> {
    ensure(
        t >= Time::ZERO,
        "Musical time must be nonnegative (durations positive)",
    )
}
fn unsupported(condition: bool, feature: &str) -> Result<()> {
    if condition {
        Ok(())
    } else {
        Err(Error::new(
            "unsupported",
            format!("Outside the D1 fixture cohort: {feature}"),
        ))
    }
}

impl Song {
    pub fn validate(&self) -> Result<()> {
        ensure(self.schema_version == 7, "Unsupported song schema version")?;
        ensure(
            identity(&self.id)
                && text(&self.title, 10000)
                && text(&self.mode, 10000)
                && self.degree_reference == "major",
            "Invalid song identity or tonality",
        )?;
        ensure(
            (10.0..=600.0).contains(&self.tempo.bpm),
            "Tempo must be between 10 and 600 BPM",
        )?;
        positive(self.tempo.beat_unit)?;
        ensure(
            text(&self.writing.instructions, 8000) && text(&self.writing.preferences, 4000),
            "Invalid writing instructions/preferences",
        )?;
        let t = &self.tables;
        ensure(t.prompts.len() <= 32, "At most 32 reusable prompts")?;
        for p in t.prompts.values() {
            ensure(
                text(&p.name, 200) && text(&p.text, 8000),
                "Invalid reusable prompt",
            )?;
        }
        // The remaining binary-coupled tables stay outside the cohort.
        unsupported(t.assets.is_empty(), "assets")?;
        unsupported(t.takes.is_empty(), "takes")?;
        let mut count = 0;
        macro_rules! table {
            ($table:ident) => {
                for (key, entity) in &t.$table {
                    ensure(
                        identity(key) && key == &entity.id && text(&entity.name, 10000),
                        &format!("Invalid identity in {}: {key}", stringify!($table)),
                    )?;
                    count += 1;
                }
            };
        }
        table!(patterns);
        table!(events);
        table!(chords);
        table!(bars);
        table!(sections);
        table!(arrangement);
        table!(parts);
        table!(voices);
        table!(occurrences);
        table!(prompts);
        table!(harmony);
        table!(markers);
        table!(phrases);
        table!(lyrics);
        table!(polyrhythms);
        table!(fretted);
        table!(fingerings);
        ensure(count <= 20000, "Song exceeds 20,000 entities")?;
        for p in t.parts.values() {
            ensure(
                ["guitar", "bass", "drums", "voice"].contains(&p.instrument.as_str())
                    && (0.0..=1.0).contains(&p.volume),
                "Invalid part",
            )?;
        }
        for v in t.voices.values() {
            ensure(
                t.parts.contains_key(&v.part_id),
                &format!("Broken parts reference: {}", v.part_id),
            )?;
        }
        for p in t.patterns.values() {
            positive(p.length)?;
            ensure(p.groups.len() <= 128, "Invalid pattern groups")?;
            let mut total = Time::ZERO;
            for &group in &p.groups {
                positive(group)?;
                total = total.checked_add(group)?;
            }
            ensure(
                p.groups.is_empty() || total == p.length,
                "Pattern groups must sum to cycle length",
            )?;
            ensure(p.length <= Time::new(10000, 1)?, "Pattern too long")?;
            if let Some(id) = &p.source_id {
                ensure(
                    t.patterns.contains_key(id),
                    &format!("Broken patterns reference: {id}"),
                )?;
                ensure(id != &p.id, "Pattern cannot vary itself")?;
            }
        }
        for c in t.chords.values() {
            pitch(&c.label_tonic)?;
            ensure(
                !c.notes.is_empty() && c.notes.len() <= 64,
                "Chord must contain 1–64 notes",
            )?;
            let mut ids = BTreeSet::new();
            for n in &c.notes {
                ensure(
                    identity(&n.id) && ids.insert(&n.id),
                    "Invalid or duplicate chord member",
                )?;
                pitch(&n.pitch)?;
            }
            ensure(
                c.label.as_ref().is_none_or(|s| text(s, 10000)),
                "Invalid chord label",
            )?;
        }
        let mut origins = BTreeSet::new();
        for e in t.events.values() {
            ensure(identity(&e.origin_id), "Invalid event origin")?;
            ensure(
                origins.insert((&e.pattern_id, &e.origin_id)),
                "Event origins must be unique within a pattern",
            )?;
            let p = t.patterns.get(&e.pattern_id).ok_or_else(|| {
                Error::new(
                    "invalid",
                    format!("Broken patterns reference: {}", e.pattern_id),
                )
            })?;
            nonnegative(e.start)?;
            positive(e.duration)?;
            ensure(
                e.start < p.length,
                "Event attack must be inside its pattern",
            )?;
            ensure(
                ["note", "chord", "drum", "rest"].contains(&e.kind.as_str())
                    && ["normal", "staccato", "sustain", "muted", "ghost"]
                        .contains(&e.articulation.as_str()),
                "Invalid event type",
            )?;
            pitch(&e.pitch)?;
            ensure(
                ["kick", "snare", "hat"].contains(&e.drum.as_str())
                    && (0.0..=1.0).contains(&e.accent),
                "Invalid event expression",
            )?;
            let chord = e.chord_id.as_ref().and_then(|id| t.chords.get(id));
            if e.kind == "chord" {
                ensure(chord.is_some(), "Broken chords reference")?;
            } else {
                ensure(e.chord_id.is_none(), "Only chord events reference chords")?;
            }
            ensure(e.performance.len() <= 64, "Invalid chord performance")?;
            ensure(
                e.kind == "chord" || e.performance.is_empty(),
                "Only chords have member performances",
            )?;
            let mut seen = BTreeSet::new();
            for perf in &e.performance {
                ensure(
                    seen.insert(&perf.member_id)
                        && chord.is_some_and(|c| c.notes.iter().any(|n| n.id == perf.member_id)),
                    "Broken/duplicate chord member reference",
                )?;
                if let Some(gain) = perf.gain {
                    ensure((0.0..=1.0).contains(&gain), "Member gain must be 0–1")?;
                }
                if let Some(articulation) = &perf.articulation {
                    ensure(
                        ["inherit", "normal", "staccato", "sustain", "muted", "ghost"]
                            .contains(&articulation.as_str()),
                        "Invalid member articulation",
                    )?;
                }
                nonnegative(perf.offset)?;
                positive(perf.duration)?;
            }
        }
        for b in t.bars.values() {
            ensure(
                t.sections.contains_key(&b.section_id),
                "Broken sections reference",
            )?;
            ensure(
                (1..=64).contains(&b.numerator)
                    && [1, 2, 4, 8, 16, 32, 64].contains(&b.denominator),
                "Invalid time signature",
            )?;
            ensure(
                !b.groups.is_empty()
                    && b.groups.iter().all(|g| (1..=64).contains(g))
                    && b.groups.iter().map(|&g| i64::from(g)).sum::<i64>()
                        == i64::from(b.numerator),
                "Beat groups must sum to numerator",
            )?;
            if let Some(actual) = b.actual {
                positive(actual)?;
                ensure(
                    actual <= Time::new(i64::from(b.numerator) * 4, i64::from(b.denominator))?,
                    "Incomplete bar cannot exceed full bar",
                )?;
            }
        }
        let mut assigned = BTreeSet::new();
        for s in t.sections.values() {
            if let Some(id) = &s.source_id {
                ensure(
                    t.sections.contains_key(id),
                    &format!("Broken sections reference: {id}"),
                )?;
                ensure(id != &s.id, "Section cannot vary itself")?;
            }
            for id in &s.bar_ids {
                ensure(
                    t.bars.get(id).is_some_and(|b| b.section_id == s.id) && assigned.insert(id),
                    "Invalid/duplicate section bar",
                )?;
            }
        }
        ensure(
            assigned.len() == t.bars.len(),
            "Every bar must appear in its section",
        )?;
        for a in t.arrangement.values() {
            ensure(
                t.sections.contains_key(&a.section_id),
                "Broken sections reference",
            )?;
        }
        let order: BTreeSet<_> = self.arrangement_order.iter().collect();
        ensure(
            order.len() == self.arrangement_order.len(),
            "Invalid arrangement order",
        )?;
        ensure(
            order.iter().all(|id| t.arrangement.contains_key(*id)),
            "Broken arrangement reference",
        )?;
        ensure(
            order.len() == t.arrangement.len(),
            "Every section occurrence needs an order",
        )?;
        for o in t.occurrences.values() {
            let p = t
                .patterns
                .get(&o.pattern_id)
                .ok_or_else(|| Error::new("invalid", "Broken patterns reference"))?;
            ensure(
                t.voices.contains_key(&o.voice_id),
                "Broken voices reference",
            )?;
            nonnegative(o.start)?;
            positive(o.span)?;
            nonnegative(o.phase)?;
            if let Some(id) = &o.section_id {
                ensure(
                    t.sections.contains_key(id),
                    &format!("Broken sections reference: {id}"),
                )?;
                let length = section_length(self, id)?;
                ensure(
                    o.start.checked_add(o.span)? <= length,
                    &format!(
                        "Placement {} exceeds its section; adjust its start/span explicitly",
                        o.name
                    ),
                )?;
            }
            ensure(o.phase < p.length, "Phase must be inside pattern")?;
            ensure(
                ["continue", "restart", "stop"].contains(&o.boundary.as_str())
                    && ["ring", "cut"].contains(&o.tails.as_str()),
                "Invalid boundary choice",
            )?;
        }
        for (table, section_id, start, duration, id) in t
            .phrases
            .values()
            .map(|e| ("phrases", &e.section_id, e.start, e.duration, &e.id))
            .chain(
                t.lyrics
                    .values()
                    .map(|e| ("lyrics", &e.section_id, e.start, e.duration, &e.id)),
            )
        {
            ensure(
                t.sections.contains_key(section_id),
                &format!("Broken sections reference: {section_id}"),
            )?;
            nonnegative(start)?;
            positive(duration)?;
            ensure(
                start.checked_add(duration)? <= section_length(self, section_id)?,
                &format!("{table}/{id} exceeds its section"),
            )?;
        }
        for l in t.lyrics.values() {
            ensure(text(&l.text, 10000), "Invalid lyric text")?;
            if let Some(id) = &l.part_id {
                ensure(
                    t.parts.contains_key(id),
                    &format!("Broken parts reference: {id}"),
                )?;
            }
            if let Some(id) = &l.phrase_id {
                let phrase = t.phrases.get(id).ok_or_else(|| {
                    Error::new("invalid", format!("Broken phrases reference: {id}"))
                })?;
                ensure(
                    l.section_id == phrase.section_id
                        && l.start >= phrase.start
                        && l.start.checked_add(l.duration)?
                            <= phrase.start.checked_add(phrase.duration)?,
                    "Linked phrase must contain its lyric span",
                )?;
            }
        }
        for section in t.sections.values() {
            let mut seen = BTreeSet::from([section.id.clone()]);
            let mut parent = section.source_id.clone();
            while let Some(id) = parent {
                ensure(!seen.contains(&id), "Cyclic sections lineage")?;
                seen.insert(id.clone());
                parent = lookup(&t.sections, "section", &id)?.source_id.clone();
            }
        }
        for pattern in t.patterns.values() {
            let mut seen = BTreeSet::from([pattern.id.clone()]);
            let mut parent = pattern.source_id.clone();
            while let Some(id) = parent {
                ensure(!seen.contains(&id), "Cyclic patterns lineage")?;
                seen.insert(id.clone());
                parent = lookup(&t.patterns, "pattern", &id)?.source_id.clone();
            }
        }
        for p in t.polyrhythms.values() {
            nonnegative(p.start)?;
            positive(p.duration)?;
            if let Some(id) = &p.section_id {
                ensure(
                    t.sections.contains_key(id),
                    &format!("Broken sections reference: {id}"),
                )?;
                ensure(
                    p.start.checked_add(p.duration)? <= section_length(self, id)?,
                    "Polyrhythm exceeds its section",
                )?;
            }
            ensure(
                (2..=8).contains(&p.lanes.len()),
                "Polyrhythm needs 2–8 lanes",
            )?;
            let mut voices = BTreeSet::new();
            for lane in &p.lanes {
                let occurrence = t.occurrences.get(&lane.occurrence_id).ok_or_else(|| {
                    Error::new(
                        "invalid",
                        format!("Broken occurrences reference: {}", lane.occurrence_id),
                    )
                })?;
                ensure((1..=64).contains(&lane.divisions), "Divisions must be 1–64")?;
                ensure(
                    occurrence.section_id == p.section_id,
                    "Polyrhythm and placements must share scope",
                )?;
                ensure(
                    voices.insert(&occurrence.voice_id),
                    "Polyrhythm lanes need distinct voices",
                )?;
            }
        }
        let mut regions: Vec<_> = t.harmony.values().collect();
        for h in &regions {
            nonnegative(h.start)?;
            positive(h.duration)?;
            pitch(&h.tonic)?;
            ensure(
                text(&h.mode, 10000) && text(&h.annotation, 10000),
                "Invalid harmonic context text",
            )?;
            if let Some(id) = &h.section_id {
                ensure(
                    t.sections.contains_key(id),
                    &format!("Broken sections reference: {id}"),
                )?;
                ensure(
                    h.start.checked_add(h.duration)? <= section_length(self, id)?,
                    "Harmonic region exceeds section",
                )?;
            }
        }
        regions.sort_by(|a, b| {
            // Matches JSON.stringify(sectionId) ordering: quoted scopes first.
            (a.section_id.is_none(), a.start).cmp(&(b.section_id.is_none(), b.start))
        });
        for pair in regions.windows(2) {
            let (a, b) = (&pair[0], &pair[1]);
            if a.section_id == b.section_id {
                ensure(
                    a.start.checked_add(a.duration)? <= b.start,
                    "Harmonic regions in the same scope cannot overlap",
                )?;
            }
        }
        for m in t.markers.values() {
            nonnegative(m.at)?;
        }
        validate_fretted(self)?;
        Ok(())
    }
}
